### PR TITLE
Setup CI for TrueNAS/Linux repository

### PR DESCRIPTION
**Pull request by umsaleem**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      - name: Install dependencies
+        run: |
+          sudo apt update
+          sudo apt install -y build-essential gawk flex bison openssl dkms \
+          libelf-dev libudev-dev libpci-dev libiberty-dev autoconf dwarves \
+          libncurses-dev libssl-dev
+
+      - name: Generate .config 
+        run: |
+          rm -rf .git .gitattributes .gitignore
```
Comment (ghost) [inline .github/workflows/ci.yml:25]: Why is this necessary?

**Review by ghost — APPROVED** [review]
I wish we could have actions run on pull request opening once, and on pushes after that, and pushes to a new branch without a pull request. It doesn't really hurt, it's just pointless to test the build twice in the same configuration. But I understand if this is a limitation of Github's actions. The truenas/zfs repo does even more actions twice, and I haven't sat down to figure out if there's a way around it yet.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          make defconfig
+          make syncconfig
+          make archprepare
+          ./scripts/kconfig/merge_config.sh .config ./scripts/package/truenas/tn.config
+          ./scripts/kconfig/merge_config.sh .config ./scripts/package/truenas/debug.config
```
Comment (ghost) [inline .github/workflows/ci.yml:30]: What do these do, merge .config into our local configs? Or the other way around?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (umsaleem) [reply .github/workflows/ci.yml:25]: Kernel makefile tries to interact with git for determining version, which results in mis-configured version and the build breaks due to that.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (umsaleem) [reply .github/workflows/ci.yml:30]: It's the other way around. We are merging tn.config and debug.config into .config.

**Comment by umsaleem** [comment]
@freqlabs [this](https://github.com/marketplace/actions/skip-duplicate-actions) seems to be helpful. Will try it out and see if it works.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (amotin) [reply .github/workflows/ci.yml:25]: I did hit the build problems myself and had to rename .git directory there and back before and after each build to still use git in between.  It is just a waste to re-download full repo for every CI run. Some sources propose to use `git clone --depth 1` at least (unless it is done already here).

**Review by amotin — APPROVED** [review]

**Comment by ghost** [comment]
Huh, one of those failed :thinking:

**Comment by umsaleem** [comment]
@freqlabs it failed while exporting the binaries. Artifact name contained '/'. 

`Artifact name is not valid: 15/merge-831cf5c0a77f8c5abceeebe311cd4992df1dc461. Contains the following character:  Forward slash /`

I have removed github.ref_name from the name now.

**Comment by ghost** [comment]
Ah so the ref name was different for the pull_request action. That makes sense.

**Review by amotin — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
We can add the skip-duplicate-actions changes separately once you get that figured out.
